common/src/address.rs

```diff
@@ -341,7 +342,69 @@ impl ShelleyAddress {
             }
         }
 
-        Ok(data)
+        data
+    }
+
+    pub fn from_bytes_key(data: &[u8]) -> Result<Self> {
+        if data.is_empty() {
+            return Err(anyhow!("empty address bytes"));
+        }
+
+        let header = data[0];
+        let network = match header & 0x0F {
+            0 => NetworkId::Testnet,
+            1 => NetworkId::Mainnet,
+            _ => return Err(anyhow!("invalid network bits in header")),
+        };
+
+        let payment_bits = (header >> 4) & 0x01;
+        let delegation_bits = (header >> 5) & 0x03;
```

Comment:
```suggestion
        let delegation_bits = (header >> 5) & 0x07;
```
Looking at CIP-0019, there are 8 valid types of Shelley addresses (2 kinds of payment part, 4 kinds of delegation part). The high bit of the header should always be 0, and if it isn't then this isn't a valid Shelley address.

This change will make the method reject header types 8 through 15 with an "invalid delegation bits" error.

Reply:
I've added a guard for `header & 0x80 != 0` in 39a8919.